internal/controller/testrun_controller.go

```diff
@@ -0,0 +1,405 @@
+	suiteRunner := &batchv1.Job{}
+	if err := r.Get(ctx, name, suiteRunner); err != nil {
+		if apierrors.IsNotFound(err) {
+			tercc, err := json.Marshal(testrun.Spec.Suites)
+			if err != nil {
+				return suiteRunner, err
+			}
+			suiteRunner = r.suiteRunnerJob(tercc, testrun)
+			if err := ctrl.SetControllerReference(testrun, suiteRunner, r.Scheme); err != nil {
+				return suiteRunner, err
+			}
+			if err := r.Create(ctx, suiteRunner); err != nil {
+				return suiteRunner, err
+			}
+		} else {
+			return suiteRunner, err
+		}
+	}
+	return suiteRunner, nil
+}
+
+// getOrCreateEnvironment attempts to get an environment or create a new one if it does not exist.
+func (r *TestRunReconciler) getOrCreateEnvironment(ctx context.Context, testrun *etosv1alpha1.TestRun, name types.NamespacedName) (*etosv1alpha1.Environment, error) {
+	environment := &etosv1alpha1.Environment{}
+	if err := r.Get(ctx, name, environment); err != nil {
+		if apierrors.IsNotFound(err) {
+			environment = &etosv1alpha1.Environment{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      make(map[string]string),
+					Annotations: make(map[string]string),
+					Name:        testrun.Name,
+					Namespace:   testrun.Namespace,
+				},
+				Spec: etosv1alpha1.EnvironmentSpec{
+					IUTProvider:            testrun.Spec.Providers.IUT,
+					ExecutionSpaceProvider: testrun.Spec.Providers.ExecutionSpace,
+					LogAreaProvider:        testrun.Spec.Providers.LogArea,
+				},
+			}
+			if err := ctrl.SetControllerReference(testrun, environment, r.Scheme); err != nil {
+				return environment, err
+			}
+			if err := r.Create(ctx, environment); err != nil {
+				return environment, err
+			}
+		} else {
+			return environment, err
+		}
+	}
+	return environment, nil
+}
+
+// isSuiteRunnerFinished checks if a suite runner has status Complete or Failed.
+func (r TestRunReconciler) isSuiteRunnerFinished(suiteRunner *batchv1.Job) (bool, batchv1.JobConditionType) {
+	if IsJobStatusConditionPresentAndEqual(suiteRunner.Status.Conditions, batchv1.JobComplete, corev1.ConditionTrue) {
+		return true, batchv1.JobComplete
+	}
+	if IsJobStatusConditionPresentAndEqual(suiteRunner.Status.Conditions, batchv1.JobFailed, corev1.ConditionTrue) {
+		return true, batchv1.JobFailed
+	}
+	return false, ""
+}
+
+// IsStatusConditionPresentAndEqual returns true when conditionType is present and equal to status.
+func IsJobStatusConditionPresentAndEqual(conditions []batchv1.JobCondition, conditionType batchv1.JobConditionType, status corev1.ConditionStatus) bool {
+	for _, condition := range conditions {
+		if condition.Type == conditionType {
+			return condition.Status == status
+		}
+	}
+	return false
+}
+
+// suiteRunnerJob is the job definition for an etos suite runner.
+func (r TestRunReconciler) suiteRunnerJob(tercc []byte, testrun *etosv1alpha1.TestRun) *batchv1.Job {
+	ttl := int32(300)
```

Comment:
Make (some of) these values configurable?

Reply:
I agree that having them hard-coded like this is sub-optimal however I do not want to provide them to the users either, which wee would do if we add it to the TestRun spec.
I suggest leaving it for now and see if we can find a better solution later.